### ISSUE TITLE
Remove dots/periods in CustomData keys

### ISSUE THREAD

**Issue by rwemay** [BHoM/Revit_Toolkit#721]
#### Description:

As discussed with @adecler , Mongo/other downstream methods don't like dots in property names/CustomData keys. Raising the issue here in Revit_TK as it's a pull from Revit that's causing the issue, but I think this issue might need to be moved to adapter(?) to do the replacing on push? Anyway, raising here initially, feel free to move/comment.

**Comment by pawelbaran** [comment]
I will keep it here for now to check whether it can't be resolved with introduction of parameter-related fragments. @alelom please feel free to transfer to the Adapter if you think it is worth resolving it there.

**Comment by alelom** [comment]
The problem with the dots arises also in the context of the UI (Explode component complaining). I had this problem too in one of the base Reflection_Engine methods and was forced to write a workaround for it in that same method. And here we talk about downstream Toolkits.

If we have to live with this issue (= working around it) I think it should be either tackled:
1) at the source (base Engine) or, less pleasantly:
2) at the level where it actually becomes problematic (the single Toolkit or UI)

I would avoid involving any other middle layer to avoid complicating this issue even further.

tl; dr:
I am generally against tackling this in the Adapter because it's not Adapter-related at all; more like a combination of UI/reflection/base framework issues.

**Comment by pawelbaran** [comment]
Right. Then again, I believe it should not be resolved at source (i.e. on pull from Revit) but where it actually occurs (Reflection, UI, push to Toolkit). Otherwise the problem will come back like a boomerang. Or am I missing a point?

**Comment by rwemay** [comment]
Should we liken this to a translation of an unsupported character? A bit like % in URL’s for spaces. So where ever there is a dot it is always replaced with something else - on Revit pull etc. Is a % or wild card even worse? Presumably replacing with 3 dots would be problematic too.

**Comment by adecler** [comment]
@alelom , the idea is that the dot is an invalid character for the BHoM in general. So it makes sense to guarantee that no CustomData key contains dots at the boundary between the BHoM and external softwares. This means inside every adapter that has a risk of containing custom data keys with dot OR as a general check in the BHoM adapter itself. The second makes more sense to me but happy to be convinced otherwise.

To clarify: This is a process purely on the keys of CustomData since the properties of BHoM object will not contain dots by definition. 

Now, as discussed with @al-fisher , external data should not be stored in CustomData but in fragments so this issue will probably become irrelevant once the migration to fragments is complete.

And then, agreed, we should have a check for dots in the UI for when user try to add CustomData to make sure no keys contain a dot.

**Comment by pawelbaran** [comment]
As discussed, for now (ever?) this issue will be resolved by introducing Revit parameter fragment on Pull (to be done together with #723).

**Comment by pawelbaran** [comment]
I am closing this one as resolved by #751 - please feel free to reopen if the issue reoccurs.
